### PR TITLE
AM-447-AM-IDP-When-using-an-internal-API-for-AM-no-validation-of-the-requests-payload-it-provided

### PR DESCRIPTION
**Pull request by farmborough**
## :id: 

Fixes [AM-447](https://gravitee.atlassian.net/browse/AM-447)

## :pencil2: 

Created a util class that checks JSON documents for negative values. 

## :memo: Test scenarios 

See the JIRA for an example IDP configuration that can be used to test this. Essentially, the steps to test are

1. Create an IDP (Suggest HTTP or In-memory as it's one of the easiest to configure)
2. Locate the IDP at the identities mongo table
3. Modify the 'configuration' field JSON to include invalid numeric values, e.g. ```"timeout" : -1000```

[AM-447]: https://gravitee.atlassian.net/browse/AM-447?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: gravitee-am-management-api/gravitee-am-management-api-service/src/main/java/io/gravitee/am/management/service/impl/utils/JsonNodeValidator.java [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+                    case BIG_INTEGER:
+                        isNegative = jsonNode.asInt() < 0;
+                        break;
+                    case FLOAT:
+                    case DOUBLE:
+                    case BIG_DECIMAL:
+                        isNegative = jsonNode.asDouble() < 0;
+                }
+                if (isNegative) {
+                    throw new IllegalArgumentException();
```
Comment (mariano-fernandez) [inline gravitee-am-management-api/gravitee-am-management-api-service/src/main/java/io/gravitee/am/management/service/impl/utils/JsonNodeValidator.java:51]: I would add a message to the exception to give some context

**Review by mariano-fernandez — CHANGES_REQUESTED** [review]

File: gravitee-am-management-api/gravitee-am-management-api-service/src/main/java/io/gravitee/am/management/service/impl/utils/JsonNodeValidator.java [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+
+    /**
+     * Validates a JSON document (in the form of a {@link JsonNode}) for 'correctness' of values.
+     * This is currently restricted to ensuring numeric values are not negative; future enhancement might involve using regex
+     * or similar to validate specific string values.
+     *
+     * @param idpConfig {@link JsonNode} representing a JSON document.
+     */
+    public static void validateConfiguration(JsonNode idpConfig) {
+        idpConfig.elements().forEachRemaining(jsonNode -> {
```
Comment (remisultan) [inline gravitee-am-management-api/gravitee-am-management-api-service/src/main/java/io/gravitee/am/management/service/impl/utils/JsonNodeValidator.java:36]: Should we handle also when a field is required ?

**Review by remisultan — COMMENTED** [review]
The validity on the IDP configuration depends on the class defined on every IDP plugin.
Shouldn't we try to serialize it with the implementation of this configuration ?

**Comment by farmborough** [comment]
> The validity on the IDP configuration depends on the class defined on every IDP plugin. Shouldn't we try to serialize it with the implementation of this configuration ?

I think what you're suggesting would be a much more thorough/better form of validation but I would argue it requires its own jira as an improvement, rather than this bug fix. I'm just validating on negative numbers but we could do with using regex or similar to provide more comprehensive validation, but that would be significantly more work.

**Review by mariano-fernandez — APPROVED** [review]

**Comment by farmborough** [comment]
Sonarcloud wasn't detecting the junit 5 test so have bypassed the quality gate.

**Pull request merged**